man/gplib.1
.TH GPLIB 1 "2007-10-27" "gputils-2.0.0_beta2" "Linux user's manual"
.SH NAME
gplib \- GNU PIC librarian
.SH SYNOPSIS
.B gplib [options] library [member]
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and
current documentation, refer to "gputils.ps" located in the gputils docs
directory.
.SH DESCRIPTION
.B gplib
creates, modifies, and extracts from COFF archives for Microchip (TM) PIC
(TM) micro-controllers.
.B gplib
is part of gputils.  Check the
.BR gputils (1)
manpage for details on other GNU PIC utilities.
.SH OPTIONS
.TP
.BR \-c ", "\-\-create
Create a new library.
.TP
.BR \-d ", "\-\-delete
Delete member from library.
.TP
.BR \-h ", "\-\-help
Show the usage message.
.TP
.BR \-n ", "\-\-no-index
Don't add symbol index.
.TP
.BR \-q ", "\-\-quiet
Quiet mode.
.TP
.BR \-r ", "\-\-replace
Add or replace member from library.
.TP
.BR \-s ", "\-\-symbols
List global symbols in library.
.TP
.BR \-t ", "\-\-list
List members in library.
.TP
.BR \-v ", "\-\-version
Show the version.
.TP
.BR \-x ", "\-\-extract
Extract member from library.
.SH SEE ALSO
.BR gputils (1)
.SH AUTHOR
Craig Franklin <craigfranklin@users.sourceforge.net>
.SH COPYRIGHT
Copyright (C) 2002, 2003, 2004, 2005 Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


history.txt
======
snutil
======
Rev:		Date:		Reason:
0.0.3		12/6/29		eval version release	
0.1.0		12/6/29		sn16 class add 4 insn (inca, jza, deca, djza)
0.2.0		12/7/3		add cod output
0.3.0		12/7/6		add prefix S2-CONFIG BIT; S3-CHIP_ID in S19 output
0.4.0		12/7/10		sn16 class add 2 insn (rra, rla)
0.5.0		12/7/13		add coff output, add env param PATH 
0.6.0		12/8/22		correct "dt" assemble
0.6.1		12/10/23	add processor MC33P94, MC33P716, MC30P011
0.7.0		12/11/19	modify memonic rsubra, rsbcra etc.
0.8.0		12/12/5		sn16 class add 2 insn (lcall, lgoto); add processor MC32P21
0.8.1		12/12/14	correct lcall/lgoto opcode
0.9.0		13/1/20		correct lcall/lgoto link
1.0.0		13/3/19		add sn13 class; comra -> comr (sn14)
1.1.0		13/3/20		adcar -> adcra (sn13)
1.2.0		13/4/12		remove website
1.3.0		13/6/19		nessary update for c compiler
1.3.1		13/7/9		add processor MC32P64
1.4.0		13/9/16		correct relocate error; expand mc30 OPF7 insn address to 7bit
1.4.1_beta	13/10/15	add sn13b class, MC34P01
1.4.2_beta	14/1/6		sn13b opcode update, MC34P01
1.5.0		14/1/13		release sn13b class, MC34P01; remove adcai of sn16 class
1.6.0		14/1/22		add sn16b modify lcall/lgoto; remove lcall/lgoto of of other class; add processor MC32P821, MC33P78, MC33P116, MC30P081
1.6.1		14/1/23		correct gplink for lcall/lgoto
1.7.0_beta	14/4/11		add sn13c for MC31P5130
1.9.1
                加入新型号7323  之前MC32P7323用的是3221 在超空间1K报错。
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.0
                加入新内核 PRO_CLASS_SN16E      for     MC32T7333、MC9902
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.2	
		add processor 4K7323 for space 4K
	MC32P7323
1.9.3
		add processor 7212 for MC32P7212
1.9.4 	
		add processor 1K6220,05K6220 for MC30P6220	17/04/17
1.9.5
		add processor 1K7041,2K7041 for MC35P7041	17/04/18
		modify 7333 setup for MC32F7333			17/04/21
1.9.6
		add processor 7122	for MC32F7122		17/04/27
1.9.7		
		add processor 9903 9904 for A380		17/05/15	
1.9.8		
		add processor 7030 for MC32P7030		17/05/25
2.0.0_beta1 17/6/19	modify cod file for macro
2.0.0_beta2 17/6/23	output section location when link error ocurrs


gputils/gplib.c
/* GNU PIC Librarian
   Copyright (C) 2001, 2002, 2003, 2004, 2005
   Craig Franklin

This file is part of gputils.
 
gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.
 
gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
 
You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "libgputils.h"
#include "gplib.h"

struct gplib_state state = {
    ar_null, /* default mode, do nothing */
    0,       /* number of objects */
};

struct symbol_table *definition_tbl = NULL;
struct symbol_table *symbol_index = NULL;

void select_mode(enum lib_modes mode)
{
  if (state.mode == ar_null)
  {
    state.mode = mode;
  }
  else
  {
    gp_error("multiple library operations selected");
  }
}

/* return the object name without the path */

static char *
object_name(char *file_name)
{
  char *name;

#ifdef HAVE_DOS_BASED_FILE_SYSTEM
  for (name = file_name + strlen(file_name) - 1; name >= file_name; --name)
  {
    if (*name == UNIX_PATH_CHAR || *name == PATH_CHAR)
    {
      return ++name;
    }
  }

  return file_name;
#else
  name = strrchr(file_name, PATH_CHAR);
  if (name)
  {
    return ++name;
  }
  else
  {
    return file_name;
  }
#endif
}

static gp_boolean
has_path(char *file_name)
{
  char *name;

  name = strrchr(file_name, PATH_CHAR);
#ifdef HAVE_DOS_BASED_FILE_SYSTEM
  if (!name)
    name = strrchr(file_name, UNIX_PATH_CHAR);
#endif

  if (name)
  {
    return true;
  }
  else
  {
    return false;
  }
}

void show_usage(void)
{
  printf("Usage: gplib [options] library [member]\n");
  printf("Options: [defaults in brackets after descriptions]\n");
  printf("  -c, --create               Create a new library.\n");
  printf("  -d, --delete               Delete member from library.\n");
  printf("  -h, --help                 Show this usage message.\n");
  printf("  -n, --no-index             Don't add symbol index.\n");
  printf("  -q, --quiet                Quiet mode.\n");
  printf("  -r, --replace              Add or replace member from library.\n");
  printf("  -s, --symbols              List global symbols in library.\n");
  printf("  -t, --list                 List members in library.\n");
  printf("  -v, --version              Show version.\n");
  printf("  -x, --extract              Extract member from library.\n");
  printf("\n");
  //printf("Report bugs to:\n");
  //printf("%s\n", PACKAGE_BUGREPORT);
  exit(0);
}

#define GET_OPTIONS "?cdhnqrstvx"

static struct option longopts[] =
    {
        {"create", 0, 0, 'c'},
        {"delete", 0, 0, 'd'},
        {"extract", 0, 0, 'x'},
        {"help", 0, 0, 'h'},
        {"no-index", 0, 0, 'n'},
        {"quiet", 0, 0, 'q'},
        {"replace", 0, 0, 'r'},
        {"symbols", 0, 0, 's'},
        {"list", 0, 0, 't'},
        {"version", 0, 0, 'v'},
        {0, 0, 0, 0}};

#define GETOPT_FUNC getopt_long(argc, argv, GET_OPTIONS, longopts, 0)

int main(int argc, char *argv[])
{
  extern int optind;
  int i = 0;
  int c;
  gp_boolean usage = false;
  gp_boolean update_archive = false;
  gp_boolean no_index = false;
  gp_archive_type *object = NULL;

  gp_init();

  /* symbols are case sensitive */
  definition_tbl = push_symbol_table(NULL, false);
  symbol_index = push_symbol_table(NULL, false);

  while ((c = GETOPT_FUNC) != EOF)
  {
    switch (c)
    {
    case '?':
    case 'h':
      usage = true;
      break;
    case 'c':
      select_mode(ar_create);
      break;
    case 'd':
      select_mode(ar_delete);
      break;
    case 'n':
      no_index = true;
      break;
    case 'q':
      gp_quiet = true;
      break;
    case 'r':
      select_mode(ar_replace);
      break;
    case 's':
      select_mode(ar_symbols);
      break;
    case 't':
      select_mode(ar_list);
      break;
    case 'v':
      fprintf(stderr, "%s\n", GPLIB_VERSION_STRING);
      exit(0);
      break;
    case 'x':
      select_mode(ar_extract);
      break;
    }
    if (usage)
      break;
  }

  if (optind < argc)
  {
    /* fetch the library name */
    state.filename = argv[optind++];
    /* some operations require object filenames or membernames */
    for (; optind < argc; optind++)
    {
      state.objectname[state.numobjects] = argv[optind];
      if (state.numobjects >= MAX_OBJ_NAMES)
      {
        gp_error("exceeded maximum number of object files");
        break;
      }
      state.numobjects++;
    }
  }
  else
  {
    usage = true;
  }

  /* User did not select an operation */
  if (state.mode == ar_null)
  {
    usage = true;
  }

  /* User did not provide object names */
  if ((state.mode != ar_list) &&
      (state.mode != ar_symbols) &&
      (state.numobjects == 0))
  {
    usage = true;
  }

  if (usage)
  {
    show_usage();
  }

  /* if we are not creating a new archive, we have to read an existing one */
  if (state.mode != ar_create)
  {
    if (gp_identify_coff_file(state.filename) != archive_file)
    {
      gp_error("\"%s\" is not a valid archive file", state.filename);
      exit(1);
    }
    else
    {
      state.archive = gp_archive_read(state.filename);
    }
  }

  /* process the option */
  i = 0;
  switch (state.mode)
  {
  case ar_create:
  case ar_replace:
    while (i < state.numobjects)
    {
      if (gp_identify_coff_file(state.objectname[i]) != object_file_v2 &&
          gp_identify_coff_file(state.objectname[i]) != object_file)
      {
        gp_error("\"%s\" is not a valid object file", state.objectname[i]);
        break;
      }
      else
      {
        state.archive = gp_archive_add_member(state.archive,
                                              state.objectname[i],
                                              object_name(state.objectname[i]));
      }
      i++;
    }
    update_archive = true;
    gp_archive_list_members(state.archive);
    break;

  case ar_delete:
    while (i < state.numobjects)
    {
      if (has_path(state.objectname[i]))
      {
        gp_error("invalid object name \"%s\"", state.objectname[i]);
        break;
      }
      object = gp_archive_find_member(state.archive, state.objectname[i]);
      if (object == NULL)
      {
        gp_error("object \"%s\" not found", state.objectname[i]);
        break;
      }
      else
      {
        state.archive = gp_archive_delete_member(state.archive,
                                                 state.objectname[i]);
      }
      i++;
    }
    update_archive = true;
    gp_archive_list_members(state.archive);
    break;

  case ar_extract:
    while (i < state.numobjects)
    {
      if (has_path(state.objectname[i]))
      {
        gp_error("invalid object name \"%s\"", state.objectname[i]);
        break;
      }
      object = gp_archive_find_member(state.archive, state.objectname[i]);
      if (object == NULL)
      {
        gp_error("object \"%s\" not found", state.objectname[i]);
        break;
      }
      else
      {
        if (gp_archive_extract_member(state.archive, state.objectname[i]))
        {
          gp_error("can't write file \"%s\"", state.objectname[i]);
          break;
        }
      }
      i++;
    }
    break;

  case ar_list:
    gp_archive_list_members(state.archive);
    break;

  case ar_symbols:
    if (gp_archive_have_index(state.archive) == 0)
    {
      gp_error("this archive has no symbol index");
    }
    else
    {
      gp_archive_read_index(symbol_index, state.archive);
      gp_archive_print_table(symbol_index);
    }
    break;

  case ar_null:
  default:
    assert(0);
  }

  /* If the archive is being modified remove the old symbol index */
  if (update_archive)
  {
    state.archive = gp_archive_remove_index(state.archive);
  }

  /* check for duplicate symbols */
  gp_archive_make_index(state.archive, definition_tbl);

  /* add the symbol index to the archive */
  if (update_archive && (!no_index))
  {
    state.archive = gp_archive_add_index(definition_tbl, state.archive);
  }

  /* write the new or modified archive */
  if (update_archive && (gp_num_errors == 0))
  {
    if (gp_archive_write(state.archive, state.filename))
      gp_error("can't write the new archive file");
  }

  if (gp_num_errors > 0)
    return EXIT_FAILURE;
  else
    return EXIT_SUCCESS;
}
